### ISSUE TITLE
Method decorated by `@singledispatchmethod` is considered incompatible override.

### ISSUE THREAD

**Issue by hongqn** [microsoft/pyright#7181]
**Describe the bug**

Pyright considers a method decorated with `@functools.singledispatchmethod` to be incompatible with the original method. It emits `reportIncompatibleMethodOverride` when overriding methods in a child class with that decorator.

**Code or Screenshots**

```python
from functools import singledispatchmethod

class Base:
    def method(self, a):
        pass

class Child(Base):
    @singledispatchmethod
    def method(self, a):
        pass

    @method.register
    def _(self, a: int):
        pass
```
Pyright v1.1.339+ reports `reportIncompatibleMethodOverride` in the [playground](https://pyright-play.net/?pythonVersion=3.11&code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlJjvAOboCmAJjpvgIbKIAWE1ybYtAUH0TommbACER1AFx8ocqLWrAonbrwAUmaumAAaKEwCUM%2BaajNRAoSOwBhNjnS11ErcdnyAAmUo16jFnZVHn5TRWVgjS0dfSMTMzkLTAFTT0jaADoQagoGZGoQDzlwqAB9TW09AylceGR3BMSbIA).  However the latest mypy does not report any error in this case, as shown in [the mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=c277f98059ec7b8affd8bf9e93801e68).

**VS Code extension or command-line**

This problem occurs since version 1.1.339 and has been tested in the pyright lsp server in neovim, pyright command line installed via homebrew, and the playground.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

You are violating the LSP (Liskov Substitution Principle) in your code sample, so a type checker should flag this as an error. If you would like to ignore the error, you can do so by adding a `# type: ignore` or `# pyright: ignore` comment. Or more generally, you can disable the `reportIncompatibleMethodOverride` diagnostic in your project if you would prefer not to see such errors.

**Comment by hongqn** [comment]
@erictraut Thank you for your reply!

However, after doing some research, I still have a question: why is it considered a violation of LSP to use the `singledispatchmethod` object instead of the `function` object in the base class? It is true that `singledispatchmethod` is not a subclass of `function` (Python does not allow inheritance from `function`), but both are descriptors, and their `__get()__` return values have the same function signature.

Since Python does not allow inheritance from `function`, using a descriptor to re-implement the base class method in a subclass to achieve the same interface (this is how it is implemented in the standard library), considering Python's duck-typing feature, can it still be considered as conforming to LSP?
